### PR TITLE
[spi_device] Revise readcmd to use cmd_info CSR

### PR DESCRIPTION
**Pull request by eunchan**
This commit revises the `spi_readcmd` submodule in the spi Flash mode.
It previously used the pre-defined opcode values and the configurations
to parse the address, the existance of the dummy cycle, read IO modes
(Single/ Dual/ Quad).

Now it uses the SW configurable cmd_info CSR to get the command
structures.

It introduces mbyte_en field too. It is not fully implemented yet. It is
a stopgap to support MByte transition. Full functionality will be
implemented later to reduce the amount of the commit size.

### CODE REVIEW

**Comment by eunchan** [comment]
Hey All, 

This is not a full functional SPI Flash read command yet. I created this PR to not have big changes in a PR. Please take a look at the cmd_info used in spi_readcmd first and ignore mbyte_en field for now.

I will create a series of PRs on top of this one. So please expect upcoming PRs looked like big changes in but ignore this commit.

File: hw/ip/spi_device/data/spi_device.hjson [hunk 1]
```diff
@@ -446,6 +446,12 @@
               If not, the address size of the command is always 3B.
               '''
           }
+          { bits:   "11"
+            name:   "mbyte_en"
```
Comment (tjaychen) [inline hw/ip/spi_device/data/spi_device.hjson:450]: sorry i can't remember, what's an `mbyte`?

Comment (eunchan) [reply hw/ip/spi_device/data/spi_device.hjson:450]: It is a byte follows address field existed in Dual IO/ Quad IO commands.

**Comment by eunchan** [comment]
Am I the only one seeing this CI error? 
image

File: hw/ip/spi_device/lint/spi_device.waiver [hunk 2]
```diff
@@ -123,6 +123,16 @@ waive -rules {NOT_USED NOT_READ} -location {spi_device.sv} \
       -regexp {'sub_(sram|p2s)_.*\[1\]' is not (used|read)} \
       -comment "CmdParse does not have SRAM intf"
 
+#### Intented Terminal States
+waive -rules {TERMINAL_STATE} -location {spi_cmdparse.sv} \
+      -regexp {StWait} \
+      -comment "CmdParse stays StWait until CSb de-assertion is received"
+
+#### Integer use
+waive -rules {INTEGER} -location {spi_cmdparse.sv} \
```
Comment (tjaychen) [inline hw/ip/spi_device/lint/spi_device.waiver:132]: i really don't understand why the tool is complaining about this..you're just indexing into an array basically right?

Comment (eunchan) [reply hw/ip/spi_device/lint/spi_device.waiver:132]: yes. It is widely used scheme. Don't know why the lint tool complains it.

**Comment by tjaychen** [comment]
> Am I the only one seeing this CI error?
> image

no i saw it earlier too. i ended up re-pushing my PR to kick things off again.

**Comment by eunchan** [comment]
> > Am I the only one seeing this CI error?
> > image
> 
> no i saw it earlier too. i ended up re-pushing my PR to kick things off again.

Hum.. Thanks. I tried multiple pushes but didn't help. Let me try again tonight.

File: hw/ip/spi_device/rtl/spi_device_pkg.sv [hunk 3]
```diff
@@ -112,6 +112,10 @@ package spi_device_pkg;
     // field.
     logic addr_4b_affected;
 
+    // Mbyte field exist. If set to 1, the command waits 1 byte before moving
+    // to dummy field. This field data is ignored.
+    logic mbyte_en;
```
Comment (tjaychen) [inline hw/ip/spi_device/rtl/spi_device_pkg.sv:117]: boy i didn't even know some of the commands did this.

Comment (eunchan) [reply hw/ip/spi_device/rtl/spi_device_pkg.sv:117]: I really don't know the usage of mbyte other than continuous read function, which basically eliminates the address field for following Read commands.

File: hw/ip/spi_device/rtl/spi_readcmd.sv [hunk 4]
```diff
@@ -318,29 +348,36 @@ module spi_readcmd
...
-    addr_latched = 1'b 0;
-
-    // TODO: handle the QuadIO case.
-    if (addr_4b_en_i) begin
-      addr_latched = (s2p_bitcnt_i == BitCntW'(8+32)) ? 1'b 1 : 1'b 0;
-    end else begin
-      addr_latched = (s2p_bitcnt_i == BitCntW'(8+24)) ? 1'b 1 : 1'b 0;
+  always_comb begin : addr_cnt
+    addr_cnt_d = addr_cnt_q;
+    if (addr_cnt_set) begin
```
Comment (tjaychen) [inline hw/ip/spi_device/rtl/spi_readcmd.sv:366]: @cfrantz pointed out yesterday there are certain commands that ignore the `addr_4b_en_i` and assume a certain size anyways. Does it make sense to expand this particular info field to be `affected_by_4b` and another bit that says `command_assumes_3b/4b` or something along those lines? Then we can natively handle that case he was referring to.

Comment (eunchan) [reply hw/ip/spi_device/rtl/spi_readcmd.sv:366]: yes. I am planning to implement that too . Let me create an issue for that and follow later

**Review by tjaychen — APPROVED** [review]
nice! i like this.
I left a couple of clarifying comments, but I don't think any of them need to be addressed in this PR.
